### ISSUE TITLE
Possible bug related to Output reuse and JavaSerializer in Kryo 2.21 / 2.22

### ISSUE THREAD

**Issue by JoshRosen** [EsotericSoftware/kryo#312]
This ticket describes a possible bug that I found in Kryo 2.21 / 2.22.  I've verified that this behavior is not present in Kryo 2.23.0, but I wanted to reach out to confirm whether this is actually a bug in Kryo or a problem in the way that I'm using it.

When is it safe to re-use an `Output` instance?  More specifically, let's say that I have a single-threaded application which creates an `Output` then repeatedly performs the following steps:

```java
for (Object obj: objects) {
  output.clear();
  kryo.reset();
  kryo.writeClassAndObject(output, obj);
  byte[] serializedBytes = output.getBytes();
  // use the same Kryo instance and a fresh Input to deserialize the object from this byte array
}
```

For this workload, should it matter whether I re-use the same `Output` and `clear()` it versus creating a fresh `Output` on each iteration?  It seems that there are cases where these two options are not equivalent.  At https://gist.github.com/JoshRosen/14ba69ef53af53ef2839, I have a ~30 line Scala program which demonstrates a corner case where this `Output` re-use is unsafe.  The problem seems to be related to Kryo's `JavaSerializer` class.

Here's the code for `JavaSerializer.write()` as of Kryo 2.21 (the version that I'm using):

```java
    private Output lastOutput;
    [...]
	public void write (Kryo kryo, Output output, Object object) {
		try {
			if (output != lastOutput) {
				objectStream = new ObjectOutputStream(output);
				lastOutput = output;
			} else
				objectStream.reset();
			objectStream.writeObject(object);
			objectStream.flush();
		} catch (Exception ex) {
			throw new KryoException("Error during Java serialization.", ex);
		}
	}
```

If we call this method two times in a row with the same `Output` then the first write will end up writing a four-byte Java serialization header.  If we call it with different instances, on the other hand, the second call will write a one-byte TC_RESET flag instead.

When we reuse the Output, all subsequent serialized objects will be missing the `OutputStream` header, leading to errors when trying to deserialize them.

This problem is not present in newer versions of Kryo.  Can you confirm whether my use of `Output` will generally be safe in newer Kryo versions, or should I be creating new `Output`s rather than reusing them?

I discovered this issue while helping out with a patch to enable Kryo serialization of closures in Apache Spark (see https://issues.apache.org/jira/browse/SPARK-7708 for more context).

**Comment by JoshRosen** [comment]
By the way, once someone confirms that this use-case is intended to work properly, I'd like to contribute a regression test to ensure that this doesn't break in the future.  Does Kryo maintain separate backport / maintenance branches for the 2.x release series?

**Comment by NathanSweet** [comment]
JavaSerializer used to rely on the Output instance. This is fixed in master. Reusing the Output is correct, you should not need to allocate a new Output.

We have tags for the 2.x releases and could commit if necessary.

**Comment by coolfrood** [comment]
@NathanSweet do you mean that the fix can be backported to 2.21? There are other dependencies (hive, chill) that make it difficult for Spark to use a newer version of Kryo.

**Comment by NathanSweet** [comment]
There is a 2.21 tag, so it should be possible to create a branch and backport fixes. Can you create a PR for that?

**Comment by coolfrood** [comment]
@NathanSweet It's been a very long time since we discussed this, but I _finally_ got around to looking at this.  I'd like to submit a PR for backporting the `JavaSerializer` fix to Kryo 2.21.  However, 2.21 is just a tag, not a branch, so I can't submit a PR against it.  Could you please create a branch for 2.21?

**Comment by magro** [comment]
@coolfrood I just created branch kryo-2.21.x from the kryo-2.21 tag, so now you should be able to submit the PR against it. I haven't really checked your changes - @NathanSweet can you have a look at it? I can do the release then.

**Comment by magro** [comment]
The fix/PR is now released as 2.21.1.
